### PR TITLE
[Diffusion] Support MOVA model

### PR DESCRIPTION
**Pull request by CloudRipple**
## Motivation

This PR integrates **MOVA** (**MO**SS **V**ideo and **A**udio Synthesis), a foundation model for synchronized video-audio generation. MOVA features an **asymmetric dual-tower architecture** with a **bidirectional cross-attention mechanism** for modality fusion, enabling high-quality multilingual lip-sync and precise audio-visual alignment.

## Modifications
**MOVA Model and VAE Configurations:**

* Added `MOVAVideoConfig` and `MOVAAudioConfig` classes for MOVA video and audio DiT models, including architecture details and parameter mappings (`mova_video.py`, `mova_audio.py`, `__init__.py`) [[1]](diffhunk://#diff-5ac48833c2247e301a6bc29f50c22659ea7a50663cd5914b7586adf3d44c4f61R1-R62) [[2]](diffhunk://#diff-68fa43ee9e4a03ac66f71a9aec36e0954082c00407ce614208a973ceb66adcb5R1-R63) [[3]](diffhunk://#diff-d34fb2f956e288da82a2eb621b3eb76282a3dc30d470006c165ed00264629181R4-R13).
* Introduced `DacVAEConfig` for the DAC audio VAE, with encoder/decoder and quantizer settings (`dac.py`, `__init__.py`) [[1]](diffhunk://#diff-c77b2a1f8c26271ec07cf560b343449c69e7cb7c8653cefa1560f8782f202012R1-R30) [[2]](diffhunk://#diff-7b823527b48db6747978b59db0508f1d5cd4efbdec7da782ca40df2414249a53R3-R8).
* Added `MOVADualTowerConfig` for dual-tower bridge model support (`mova_dual_tower.py`, `__init__.py`) [[1]](diffhunk://#diff-6326f77feae838ad3bf4ab653d18ba7d629ae08c00e960246a17a9015f3da3ceR1-R50) [[2]](diffhunk://#diff-ec9fe9e4d0044ba268fce88756ce161f99094208cb7100d19ff5cc335beb85dfR1-R7).

**Pipeline and Sampling Configuration:**

* Implemented `MOVAPipelineConfig` for the MOVA pipeline, handling preprocessing, latent shape calculation, and normalization/denormalization for video and audio, with 360P and 720P variants (`mova.py`, `__init__.py`) [[1]](diffhunk://#diff-567fea1153b785a876b0a2e601b8cecd1cfcff267da011674103188fd0d6ca7fR1-R188) [[2]](diffhunk://#diff-dc8a2d033074b69799d9427cba1a0ef3a1f6a210ee5f6f70be7571280d574259R22) [[3]](diffhunk://#diff-dc8a2d033074b69799d9427cba1a0ef3a1f6a210ee5f6f70be7571280d574259R42).
* Added `MOVASamplingParams`, `MOVA_360P_SamplingParams`, and `MOVA_720P_SamplingParams` for MOVA-specific sampling parameters and supported resolutions (`mova.py`).

**Integration and Registry Updates:**

* Registered MOVA pipeline and sampling configs in the main registry, enabling automatic selection based on model identifiers (`registry.py`) [[1]](diffhunk://#diff-8da1a40e0b4fd58498b3c3d417bcb0f391766c8d4581217a086ed23494d03967R48-R51) [[2]](diffhunk://#diff-8da1a40e0b4fd58498b3c3d417bcb0f391766c8d4581217a086ed23494d03967R77-R80) [[3]](diffhunk://#diff-8da1a40e0b4fd58498b3c3d417bcb0f391766c8d4581217a086ed23494d03967R542-R552).

**Runtime Enhancements:**

* Updated the runtime generator to handle audio output and sample rate in the result objects, supporting MOVA's video+audio outputs (`diffusion_generator.py`) [[1]](diffhunk://#diff-6d8020638f7ff5af6eae0436b27808eb02b3e389dd943c4b750fab45ce6ecf96R236-R254) [[2]](diffhunk://#diff-6d8020638f7ff5af6eae0436b27808eb02b3e389dd943c4b750fab45ce6ecf96R264-R265).

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/model_stages/mova.py [hunk 1]
```diff
@@ -0,0 +1,1041 @@
...
+                        emitter.close()
+                    except Exception:
+                        pass
+                nvtx_emitters = []
+
+                try:
+                    nsys_nvtx_context.__exit__(None, None, None)
+                finally:
+                    nsys_nvtx_context = None
+
```
[the remark this answers is not in the document]
Comment (cms42) [reply python/sglang/multimodal_gen/runtime/models/model_stages/mova.py]: Removed those profiling logic. Not planned for public release.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/input_validation.py [hunk 2]
```diff
@@ -175,6 +176,36 @@ def preprocess_condition_image(
...
+            config = server_args.pipeline_config
+            image, (final_w, final_h) = (
+                server_args.pipeline_config.preprocess_condition_image(
+                    image, width, height, self.vae_image_processor
+                )
+            )
+            batch.condition_image = image
+            batch.width = final_w
+            batch.height = final_h
+
```
[the remark this answers is not in the document]
Comment (cms42) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/input_validation.py:229]: Resolved. Wrote a new helper function and moved the logics there.

File: python/sglang/multimodal_gen/configs/models/bridges/mova_dual_tower.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        default_factory=lambda: [_is_conditioner_block]
+    )
+
+    # Key mapping: HF weights -> SGLang model weights
+    # The weights in safetensors use the same naming convention as our model,
+    # so minimal mapping is needed. The main changes are for MLP layers
+    # where we use fc_in/fc_out naming.
+    param_names_mapping: dict = field(default_factory=dict)
+    reverse_param_names_mapping: dict = field(default_factory=dict)
+    lora_param_names_mapping: dict = field(default_factory=dict)
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/models/bridges/mova_dual_tower.py]: do we need these?

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py [hunk 4]
```diff
@@ -844,6 +983,18 @@ def load_customized(
         return model
 
 
+class AudioTransformerLoader(BaseTransformerLoader):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/loader/component_loader.py]: could we avoid adding new classes, but pass values instead?

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py [hunk 5]
```diff
@@ -268,18 +268,32 @@ def for_module_type(
...
-
         # Loaders for audio/video specific components that might vary
         av_module_loaders = {
-            "audio_vae": (VAELoader, "diffusers"),
-            "vocoder": (VocoderLoader, "diffusers"),
+            "audio_dit": (AudioTransformerLoader, "diffusers"),
+            "audio_vae": (AudioVAELoader, "diffusers"),
             "connectors": (AdapterLoader, "diffusers"),
+            "dual_tower_bridge": (BridgeLoader, "diffusers"),
+            "video_dit": (TransformerLoader, "diffusers"),
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/loader/component_loader.py:277]: there's probably no need to introduce a new loader for audio at the moment

File: python/sglang/multimodal_gen/runtime/models/bridges/mova_dual_tower.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+@torch.compile(fullgraph=True)
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/bridges/mova_dual_tower.py]: we have a `apply_flashinfer_rope_qk_inplace`

File: python/sglang/multimodal_gen/registry.py [hunk 7]
```diff
@@ -543,6 +551,17 @@ def _register_configs():
             "FastVideo/FastWan2.1-T2V-1.3B-Diffusers",
         ],
     )
+    # MOVA
+    register_configs(
+        sampling_param_cls=MOVA_360P_SamplingParams,
+        pipeline_config_cls=MOVA360PConfig,
+        model_detectors=[lambda hf_id: "360p" in hf_id.lower()],
```
Comment (mickqian) [inline python/sglang/multimodal_gen/registry.py]: this is way too broad

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/registry.py (hunk 7 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/registry.py]: > this is way too broad

Where is the `hf_id` field being checked in `model_detectors` obtained from?

File: python/sglang/multimodal_gen/configs/models/bridges/mova_dual_tower.py (hunk 3 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/configs/models/bridges/mova_dual_tower.py]: removed.

File: python/sglang/multimodal_gen/registry.py (hunk 7 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/registry.py]: Added 'mova' identifier to model_detectors

File: python/sglang/multimodal_gen/runtime/models/bridges/mova_dual_tower.py (hunk 6 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/runtime/models/bridges/mova_dual_tower.py]: Applied. Seeing a 3% speedup per step on average, with zero precision loss compared to the baseline.

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py (hunk 4 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/runtime/loader/component_loader.py]: Fixed. Please review again.

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py (hunk 5 above)
Comment (CloudRipple) [reply python/sglang/multimodal_gen/runtime/loader/component_loader.py:277]: Fixed. Please review again.

**Review by yhyang201 — APPROVED** [review]

**Review by mickqian — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/models/schedulers/flow_match_pair.py [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+            raise TypeError("pair_postprocess 必须是可调用对象或 None")
+        self._pair_postprocess_fn = fn
+        self._pair_postprocess_requires_source = (
+            False if fn is None else bool(getattr(fn, "_requires_source", False))
+        )
+        if self.timesteps is None or self.sigmas is None:
+            raise RuntimeError("调度器未初始化，请先调用 set_timesteps()")
+        self._refresh_pair_cache()
+
+    def set_pair_postprocess_by_name(self, name: str | None, **kwargs):
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/schedulers/flow_match_pair.py:222]: Please clean code
